lib/differ.rb

```diff
@@ -0,0 +1,63 @@
+module Differ
+  class << self
+    def get_dmp_diffs before, after
+      dmp = DiffMatchPatch.new
+      # https://github.com/google/diff-match-patch/wiki/API#diff_maintext1-text2--diffs
+      diffs = dmp.diff_main(before, after)
+      # https://github.com/google/diff-match-patch/wiki/API#diff_cleanupsemanticdiffs--null
+      dmp.diff_cleanupSemantic(diffs)
+      diffs
+    end
+
+    def calculate_ranges dmp_diffs
+      dmp_diffs.reduce([]) do |diffs_with_ranges, diff|
+        start_before = diffs_with_ranges.last&.[](0) == :insert ?
+                         diffs_with_ranges.last[2].min :
+                         diffs_with_ranges.last&.[](2)&.max || 0
+
+        start_after = diffs_with_ranges.last&.[](0) == :delete ?
+                        diffs_with_ranges.last[3].min :
+                        diffs_with_ranges.last&.[](3)&.max || 0
+
+        diffs_with_ranges + [diff + [start_before..start_before + diff[1].length,
+                                     start_after..start_after + diff[1].length]]
+      end
+    end
+
+    # return structure:
+    # [:action, "affected string", (before..range), (after..range)]
+    def get_diffs before, after
+      calculate_ranges(get_dmp_diffs(before, after))
+    end
+
+    def get_first_delta_offset diffs
+      diffs.find { |diff| diff[0] != :equal }&.[](2)&.min || -1
+    end
+
+    def range_was_deleted? diffs, range
+      !!diffs.find { |diff|
+        diff[0] == :delete &&
+          diff[2].min <= range.min &&
+          diff[2].max >= range.max
+      }
+    end
+
+    def get_delta_at_offset diffs, offset
```

Comment:
Yeah, that's not quite accurate. The comment mentions that we reverse the list because we want the last *diff*. Diffs are often a paired in the instance of a delete followed by an insert, both of which would match for a given offset because they're part of what's essentially a single action at a single offset in the text. We want the insert (i.e. the "last" diff) because that contains the information about how the text changed.